### PR TITLE
[Clang][Parser] Fix crash of clang when using C++ constructs like :: in C code

### PR DESCRIPTION
**Pull request by ChipsSpectre**
As stated in this issue, https://github.com/llvm/llvm-project/issues/73559, clang (C version of clang, not C++) crashes when C++ components are parsed.

Even a file which only contains a double-colon "::" leads to this kind of crash.

The reason for this behavior is that TryAnnotateCXXScopeToken() of the clang parser assumes to be executed in C++ mode only. 

But in 2 places in the code, there was no guard to ensure this function is only called in C++ mode. When it is called nonetheless, the aforementioned error is created.

### CODE REVIEW

File: clang/lib/Parse/ParseDeclCXX.cpp
Comment (Fznamznon) [inline clang/lib/Parse/ParseDeclCXX.cpp:2705]: The caller is named `ParseCXXClassMemberDeclaration` so, I'm not sure it can and should be called outside of C++. Perhaps we should assert instead of checking that is its C++.

**Review by Fznamznon — COMMENTED** [review]
Could you please add a test, somewhere in `clang/test/Parser` (there is also a lot of other tests showing how it is done) and a release note to `clang/docs/ReleaseNotes.rst`, perhaps at the end of the list under `Bug Fixes in This Version` heading.

**Comment by ChipsSpectre** [comment]
The pull request is updated now, and passing tests and formatting checks.

Is it ready to be merged in its new state?

File: clang/docs/ReleaseNotes.rst
Comment (Fznamznon) [inline clang/docs/ReleaseNotes.rst:644]:
```suggestion
- Fix crash when using C++ only tokens like ``::`` in C compiler clang.
```

File: clang/test/Parser/cxx-in-c.c
Comment (Fznamznon) [inline clang/test/Parser/cxx-in-c.c:6]:
```suggestion

// GH73559
::; // expected-error{{expected identifier or '('}}
```

**Review by Fznamznon — COMMENTED** [review]
The PR can be landed after a formal approval. Please see https://llvm.org/docs/CodeReview.html#lgtm-how-a-patch-is-accepted .

I gave a couple of NITs otherwise the patch looks ok to me. I'm a bit new here, so someone else should approve. I added more folks to reviewers list.

**Comment by ChipsSpectre** [comment]
The tests also pass after the review comments, I just validated on my machine.

**Comment by ChipsSpectre** [comment]
@cor3ntin could you please give your review?
@shafik or would you have time to do this?

The fix is ready to be merged, and all review comments are addressed.

**Comment by antoniofrighetto** [comment]
Could you kindly squash everything into one commit and provide a meaningful git commit title and git message description? For example:
```
[clang][Parse] `TryAnnotateCXXScopeToken` to be called only when parsing C++

Assume `TryAnnotateCXXScopeToken` to be parsing C++ code in all of its paths.

Fixes: https://github.com/llvm/llvm-project/issues/73559.
```

**Comment by ChipsSpectre** [comment]
The changes are squashed into one commit now.

Sorry for the force push, it was needed to rebase on main.

Do you think it can be merged now?

File: clang/docs/ReleaseNotes.rst
Comment (shafik) [inline clang/docs/ReleaseNotes.rst:644]: It looks like you added your note in the middle of another note that has a before and after description of the bug they fixed.

Please make sure you move this to the end of the list. 

After that I think this is good.

**Comment by shafik** [comment]
In the future please add a more description title to your PR. Your summary is great though.

Also in this case this is a clang and parser fix so you should prefix your title with `[Clang][Parser]`

**Comment by ChipsSpectre** [comment]
@shafik ok, done

**Comment by tbaederr** [comment]
There's a conflict in the `ReleaseNotes.rst` again unfortunately :(

**Comment by ChipsSpectre** [comment]
The Fork is now rebased on main, so the conflict in `ReleaseNotes.rst` is resolved.

@tbaederr Do you think it is ready to be merged now?

File: clang/lib/Parse/ParseDeclCXX.cpp
Comment (cor3ntin) [inline clang/lib/Parse/ParseDeclCXX.cpp:2683]:
```suggestion
         "ParseCXXClassMemberDeclaration should only be called in C++ mode");
```

**Review by cor3ntin — COMMENTED** [review]
Thanks for fixing this bug.
LGTM modulo nit

**Comment by ChipsSpectre** [comment]
ok @cor3ntin, the nit is addressed.

Is the pull request now ready to be merged?

**Review by cor3ntin — APPROVED** [review]
Thanks for working on this! It looks good now

**Comment by ChipsSpectre** [comment]
@cor3ntin thank you for your fast review!

Is it possible to merge it after your approval, or is it needed to wait for all reviewers?

**Review by shafik — APPROVED** [review]
If you have addressed all review comments and no one has asked you to wait to land the PR then you can land once you get an approval.

**Comment by ChipsSpectre** [comment]
@tahonermann @AaronBallman @nicolasvasilache @dcaballe Could you please approve this issue?

The bug is fixed and already approved twice, but the pull request can not be landed without your approval.

**Comment by tbaederr** [comment]
It doesn't need a specific approval, just _any_. And @shafik has already approved this PR, so it can now be merged.

**Comment by tbaederr** [comment]
The build fails though, with this error:

``
tools/clang/docs/ReleaseNotes.rst:691:Unknown target name: "#73559 https://github.com/llvm/llvm-project/issues/73559"
``

**Comment by ChipsSpectre** [comment]
@tbaederr The build is fixed now. Can you merge the pull request?

Since I have no write access to llvm-project, it seems I can not do it on my own:

Screenshot

**Comment by tbaederr** [comment]
The documentation build still fails fwiw.

**Comment by ChipsSpectre** [comment]
Sorry @tbaederr, I made a mistake in fixing the ReleaseNotes.

Now I added the missing brackets to the link for this issue in the ReleaseNotes.

Could you merge the pull request now?

**Review by cor3ntin — APPROVED** [review]

**Comment by cor3ntin** [comment]
Thanks for working on that!

**Pull request merged**
